tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Wcf/BeforeSendReplyIntegration.cs

```diff
@@ -0,0 +1,75 @@
+    /// </summary>
+    [InstrumentMethod(
+        AssemblyName = "System.ServiceModel",
+        TypeName = "System.ServiceModel.Dispatcher.ImmutableDispatchRuntime",
+        MethodName = "BeforeSendReply",
+        ReturnTypeName = ClrNames.Void,
+        ParameterTypeNames = new[] { "System.ServiceModel.Dispatcher.MessageRpc&", $"{ClrNames.Exception}&", $"{ClrNames.Bool}&" },
+        MinimumVersion = "4.0.0",
+        MaximumVersion = "4.*.*",
+        IntegrationName = WcfCommon.IntegrationName)]
+    [Browsable(false)]
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public class BeforeSendReplyIntegration
+    {
+        /// <summary>
+        /// OnMethodBegin callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <typeparam name="TMessageRpc">Type of the rpc message</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <param name="rpc">MessageRpc instance</param>
+        /// <param name="exception">The exception instance from preparing the message reply</param>
+        /// <param name="thereIsAnUnhandledException">A flag indicating whether exception caught from preparing the message reply is unhandled</param>
+        /// <returns>Calltarget state value</returns>
+        internal static CallTargetState OnMethodBegin<TTarget, TMessageRpc>(TTarget instance, ref TMessageRpc rpc, ref Exception exception, ref bool thereIsAnUnhandledException)
+        {
+            if (!Tracer.Instance.Settings.IsIntegrationEnabled(WcfCommon.IntegrationId) || !Tracer.Instance.Settings.DelayWcfInstrumentationEnabled || WcfCommon.GetCurrentOperationContext is null)
+            {
+                return CallTargetState.GetDefault();
+            }
+
+            var rpcProxy = rpc.DuckCast<MessageRpcStruct>();
+            if (((IDuckType?)rpcProxy.OperationContext.RequestContext)?.Instance is object requestContextInstance
+                && WcfCommon.Scopes.TryGetValue(requestContextInstance, out var scope))
+            {
+                return new CallTargetState(scope);
+            }
+
+            return CallTargetState.GetDefault();
+        }
+
+        /// <summary>
+        /// OnMethodEnd callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <param name="exception">Exception instance in case the original code threw an exception.</param>
+        /// <param name="state">Calltarget state value</param>
+        /// <returns>A response value, in an async scenario will be T of Task of T</returns>
+        internal static CallTargetReturn OnMethodEnd<TTarget>(TTarget instance, Exception exception, in CallTargetState state)
+        {
+            state.Scope.DisposeWithException(exception);
```

Comment:
I wonder if we should use the `ref Exception` argument. BeforeSendReply doesn't throw:

https://referencesource.microsoft.com/#system.servicemodel/System/ServiceModel/Dispatcher/ImmutableDispatchRuntime.cs,347
```

if (exception == null)
{
	exception = e;
}
thereIsAnUnhandledException = (!this.error.HandleError(e)) || thereIsAnUnhandledException;
```

But the snapshots are fine so I don't know

Reply:
Oh, I think it's only the exceptions thrown by the message inspector

Reply:
Yep this exception argument is related to the message inspectors and not the endpoint so that's why I didn't use it